teps/0123-specifying-on-demand-retry-in-pipelinetask.md

```diff
@@ -0,0 +1,183 @@
+---
+status: proposed 
+title: Specifying on-demand-retry in a PipelineTask
+creation-date: '2022-09-16'
+last-updated: '2022-10-11'
+authors:
+- '@pritidesai'
+---
+
+# TEP-0123: Specifying on-demand-retry in a PipelineTask
+
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Requirements](#requirements)
+  - [Use Cases](#use-cases)
+    - [CD Use Case](#cd-use-case)
+    - [CI Use Case](#ci-use-case)
+- [References](#references)
+<!-- /toc -->
+
+## Summary
+
+This TEP proposes a mechanism to allow users to choose when to retry a failed `pipelineTask` of a pipeline.
+This kind of `on-demand` retry of a `pipelineTask` is executed with the original `params` and `specifications`.
+The `on-demand` retry of a `pipelineTask` allow users to continue to retry until the task succeeds. After the offending task
+succeeds, the rest of the `pipeline` continues executing as usual until the completion. This retry of a `pipelineTask`
+just like the existing `retry` strategy is only applicable within the context of an original `pipelineRun` i.e.
+a `pipelineTask` can only be retried until the `pipelineRun` that it's part of is `running`.
+
+## Motivation
+
+The primary motivation of this proposal is to enable pipeline authors to overcome any flakes without re-running
+the whole pipeline but to re-run the flakes and continue the execution until the pipeline finishes. The main motivation
+of this proposal is to enable pipeline authors to specify **when** to re-run such flakes.
+
+Tekton controller does support [retries](https://github.com/tektoncd/pipeline/blob/main/docs/pipelines.md#using-the-retries-field)
+in which a pipeline author can specify the number of the times a task can be retried when it fails. The `retries` are designed
+such that once a task fails, it is retried immediately. Other CI/CD systems supports an additional retry strategies, such as:
+
+* Argo Workflow offers a variety of retry policies. One of them is [Back-Off](https://argoproj.github.io/argo-workflows/retries/#back-off),
+in which a user can configure the delay between retries.
+
+* GitHub action supports a retry of an action allowing to set some delay between attempts through [attempt_delay](https://github.com/marketplace/actions/retry-action#attempt_delay).
+
+These additional retry strategies allows users to specify a reasonable delay during which a transient failure can be
+fixed.
+
+Jenkins takes `retry` one step further where a user can specify [conditions](https://www.jenkins.io/doc/pipeline/steps/workflow-basic-steps/#retry-retry-the-body-up-to-n-times)
```

Comment:
Right, retry logic is not on demand. Jenkins is responsible for retrying in case of a connection issue.

But, Jenkins does support catching failures and further processing somehow:

"Retry the block (up to N times) if any exception happens during its body execution. If an exception happens on the final attempt then it will lead to aborting the build (unless it is caught and processed somehow). User aborts of the build are not caught."

I have not looked into it further though.

Also, Jenkins support `conditions` which is very advanced in terms of specifying what condition triggers the retry. Also, Jenkins retry logic understands the difference between infrastructure failure (connection to agent itself is broken) v/s build failure and allows pipeline author to choose to retry in case of an infrastructure failure. Jenkins retry logic can also distinguishes between pod termination v/s pod deletion.


"conditions (optional)
Conditions under which the block should be retried. If none match, the block will fail. If there are no specified conditions, the block will always be retried except in case of user aborts.
Array / List of Nested Choice of Objects
agent
Detects that a node block, or certain steps inside it such as sh, failed for reasons which are likely due to infrastructure rather than the behavior of the build. If the connection to an agent is broken or the agent is removed from the list of executors while in use (typically in response to the disappearance of underlying cloud resources), this condition will allow retry to allocate a fresh agent and try the whole block again.
kubernetesAgent
Similar to agent() (Agent errors) but tailored to agents provisioned from a Kubernetes cloud. Unlike the generic agent error condition, this will ignore certain pod termination reasons which are likely to be under the control of the Pipeline author (e.g., OOMKilled) while still allowing retry to recover after common cases of pod deletion.
handleNonKubernetes : boolean (optional)
Behave like the generic agent() (Agent errors) when applied to a non-Kubernetes agent. Useful in cases where it is hard to predict in a job definition whether a Kubernetes or other sort of agent will be used.
nonresumable
The Jenkins controller was restarted while the build was running a step which cannot be resumed. Some steps like sh or input are written to survive a Jenkins restart and simply pick up where they left off when the build resumes. Others like checkout or junit normally complete promptly but cannot tolerate a restart. In case one of these steps happened to be in progress when Jenkins shut down, the resumed build will throw an error; using this condition with retry allows the step (or the whole enclosing node block) to be rerun."

What we are trying build in this PR is the limitation in Jenkins that the users are experiencing: https://issues.jenkins.io/browse/JENKINS-54735 And something on their roadmap.